providers/implementations/keymgmt/ecx_kmgmt.c

```diff
@@ -588,6 +592,73 @@ static const OSSL_PARAM *ecx_gen_settable_params(ossl_unused void *genctx,
     return settable;
 }
 
+static int ecd_keygen_pairwise_test(ECX_KEY *ecx) {
+    int ret = 0;
+    OSSL_SELF_TEST *st = NULL;
+    OSSL_CALLBACK *cb = NULL;
+    void *cbarg = NULL;
+
+    unsigned char msg[16] = {0};
+    size_t msg_len = sizeof(msg);
+    unsigned char sig[ED448_SIGSIZE] = {0};
+    size_t sig_len = sizeof(sig);
+
+    EVP_SIGNATURE *alg = NULL;
+    void *provctx = NULL;
+    void *algctx = NULL;
+    const char *instance = (ecx->type == ECX_KEY_TYPE_ED25519) ?
+                           "Ed25519" : "Ed448";
+    const OSSL_PARAM params[] = {
+        OSSL_PARAM_utf8_string(OSSL_SIGNATURE_PARAM_INSTANCE, instance,
+                               strlen(instance)),
+        OSSL_PARAM_END
+    };
+
+    OSSL_SELF_TEST_get_callback(ecx->libctx, &cb, &cbarg);
+
+    st = OSSL_SELF_TEST_new(cb, cbarg);
+    if (st == NULL)
+        return 0;
+
+    OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_PCT,
+                           OSSL_SELF_TEST_DESC_PCT_EDDSA);
+
+    alg = EVP_SIGNATURE_fetch(ecx->libctx, instance, NULL);
```

Comment:
If you look at other places that do this, they all avoid the overhead of doing the fetch by calling directly into low level API's. Its bad enough that we need to do this in the first place :). So crypto/ecx.h would be better to use for this.
Could you also add a comment block to the top of the function that references the relevant section in the FIPS 140-3 IG that forces us to do this..

Reply:
Does it need to do sign + verify or is just doing a KAT sufficient for this purpose?

Reply:
> If you look at other places that do this, they all avoid the overhead of doing the fetch by calling directly into low level API's. Its bad enough that we need to do this in the first place :). So crypto/ecx.h would be better to use for this.

Yes, I noticed that RSA and ECDSA use low-level API like `ECDSA_do_sign` to do this, but didn't aware there are `ossl_ed25519_sign` and other APIs for EDDSA.

I will try to change the implementation using them. Thank you.

> Could you also add a comment block to the top of the function that references the relevant section in the FIPS 140-3 IG that forces us to do this..

Of course, I will add it in next commit.

> Does it need to do sign + verify or is just doing a KAT sufficient for this purpose?

The addition comment in IG states

```
Though not a CAST, a pairwise consistency test (PCT) shall be conducted for every generated
public and private key pair for the applicable approved algorithm ... shall be performed.

If at the time a PCT on a key pair is performed it is known whether the keys will be used in a key
agreement scheme, digital signature algorithm or to perform a key transport, then the PCT shall be
performed consistent with the intended use of the keys ..., even if the underlying standard does not
require a PCT.

... The PCT shall be performed either when keys are generated/imported, prior to the first
exportation, or prior to the first operational use (if not exported before the first use).
```

In my opinion, I think it is needed to do sign + verify for every generated key, just like DSA, ECDSA and RSA.

If there is a worry about it, I may ask our labs to confirm it.

Reply:
As we cant use X25519/X448 in FIPS 140-3 currently, then we know that the keygen result can only be used  for signature purposes.. Until that changes we can only do the sign test here :)

Reply:
The question is whether this has to be part of the key generation algorithm or if it would be sufficient to make mandatory that an application uses EVP_PKEY_pairwise_check() before the first use of the keypair. However please note that EVP_PKEY_pairwise_check() does not do a PCT for EdDSA keys - it just regenerates the public key from the private key and compares the result with the original public key. So we would have to change the pairwise check anyway.

Reply:
> The question is whether this has to be part of the key generation algorithm or if it would be sufficient to make mandatory that an application uses EVP_PKEY_pairwise_check() before the first use of the keypair.

After confirming with our lab, the lab states that PCT should be performed automatically as soon as the key is generated. Therefore, it should be a part of the key generation.

> So we would have to change the pairwise check anyway.

Please let me confirm it again, so I should replace callings of `ossl_ed*_public_from_private` with `ecd_keygen_pairwise_test` (maybe rename to a more common name) in `ecx_key_pairwise_check` for types `ECX_KEY_TYPE_ED25519` and `ECX_KEY_TYPE_ED448`. Is it right?

Reply:
> Please let me confirm it again, so I should replace callings of `ossl_ed*_public_from_private` with `ecd_keygen_pairwise_test` (maybe rename to a more common name) in `ecx_key_pairwise_check` for types `ECX_KEY_TYPE_ED25519` and `ECX_KEY_TYPE_ED448`. Is it right?

I would do that at least within the FIPS_MODULE ifdefs. @slontis What is your opinion? IMO the current pairwise check for EdDSA keys is not FIPS compliant.

Reply:
The commit pushed.

I make it run sign + verify operations in FIPS provider and original operation in default provider.